hammerblade/torch/kernel/hb_tiled_for.hpp

```diff
@@ -74,218 +74,121 @@ inline void calc_range(hb_range* range, size_t numel) {
-    // iterating over all elementes
-    //-----------------------------
-    hb_range range;
-    calc_range(&range, res.numel());
-    size_t start = range.start;
-    size_t end   = range.end;
-
-    for (size_t idx = start; idx < end; idx++) {
-      scalar_t* res_dp = (data[0] + offset_calc(idx, res));
-      scalar_t* input_dp = (data[1] + offset_calc(idx, input));
-      scalar_t* tensor1_dp = (data[2] + offset_calc(idx, tensor1));
-      scalar_t* tensor2_dp = (data[3] + offset_calc(idx, tensor2));
-      *res_dp = functor(*input_dp, *tensor1_dp, *tensor2_dp);
+    UNROLL(16) for (size_t idx = start; idx < end; idx++) {
+      res_ptr[offset_calc(idx, res)] =
+        functor();
     }
   }
 }
 
-// =========================================================
-// Tile Element-wise for -- Binary ops
-//
-// This function calculates the per tile range automatically
-//==========================================================
-
-template<typename scalar_t, typename F>
-inline void hb_tiled_foreach(HBTensor<scalar_t> res,
-                               HBTensor<scalar_t> input,
-                               HBTensor<scalar_t> other,
-                               F functor) {
-  scalar_t* data[3];
-  data[0] = (scalar_t*)res.data_ptr();
-  data[1] = (scalar_t*)input.data_ptr();
-  data[2] = (scalar_t*)other.data_ptr();
-
+// Unary
+template<typename scalar_t, typename F, typename... P>
+__attribute__((noinline)) void hb_tiled_foreach_impl(
+      size_t start, size_t end, F functor,
+      HBTensor<scalar_t> res,
+      HBTensor<scalar_t> tensor_arg0,
+      __remote scalar_t* NOALIAS res_ptr,
+      __remote scalar_t* NOALIAS tensor_data_ptr0) {
   // is_trivial_1d
   if(res.ndim() == 1) {
-
-    //-----------------------------
-    // collect metadata
-    //-----------------------------
-    uint32_t strides[3];
-    strides[0] = (res.get_strides())[0];
-    strides[1] = (input.get_strides())[0];
-    strides[2] = (other.get_strides())[0];
-
-    //------------------------------
-    // in the case where stride is 0
-    //------------------------------
-    scalar_t fixed_data[3];
-    for (size_t i = 0; i < 3; i++) {
-      if (strides[i] == 0) {
-        fixed_data[i] = *data[i];
-        data[i] = &fixed_data[i];
-      }
-    }
-
-    //-----------------------------
-    // iterating over all elementes
-    //-----------------------------
-    hb_range range;
-    calc_range(&range, res.numel());
-    size_t start = range.start;
-    size_t end   = range.end;
-
-    data[0] += strides[0] * start;
-    data[1] += strides[1] * start;
-    data[2] += strides[2] * start;
-    for (size_t idx = start; idx < end; idx++) {
-      *data[0] = functor(*data[1], *data[2]);
-      data[0] += strides[0];
-      data[1] += strides[1];
-      data[2] += strides[2];
+    UNROLL(16) for(size_t idx = start; idx < end; idx++) {
+      res_ptr[idx * res.get_strides()[0]] =
+        functor(tensor_data_ptr0[idx * tensor_arg0.get_strides()[0]]);
```

Comment:
How does the assembly look like? does it have the multiplication in the inner loop? I guess no?

Reply:
No, they are compiled to a series of adds... felt `data[idx * stride]` format is easier to read.